src/PureScript/Backend/Optimizer/Semantics.purs

```diff
@@ -30,6 +33,26 @@ type Spine a = Array a
 
 type RecSpine a = NonEmptyArray (Tuple Ident (Lazy a))
 
+data Floatable a = FloatMe BackendSemantics (BackendSemantics -> a)
+
+derive instance Functor Floatable
+
+type FreeFloatableWithEnv = Free Floatable (Env -> BackendSemantics)
+type FreeFloatable = Free Floatable BackendSemantics
+
+floatMe :: BackendSemantics -> FreeFloatable
+floatMe a = liftF $ FloatMe a identity
+
+runFloatLetsWithEnv :: Boolean -> Env -> FreeFloatableWithEnv -> BackendSemantics
+runFloatLetsWithEnv atTop ee ff = go ff ee
+  where
+  go = resume >>> case _ of
+    Left (FloatMe sem f) -> \e1 -> if atTop then go (f sem) e1 else guardFail sem \sem' -> evalAssocLet e1 sem' \e2 v -> go (f v) e2
```

Comment:
evalAssocLet already guards on Fail, so the guardFail check here is unnecessary.